internal/controller/configmap.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/mitchellh/mapstructure"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+)
+
+type OnDemandTLS struct {
+	Ask      string          `json:"ask"`
+	Interval caddy2.Duration `json:"interval"`
+	Burst    int             `json:"burst"`
+}
+
+type ConfigMapOptions struct {
+	Debug  bool   `json:"debug"`
+	AcmeCA string `json:"acme-ca"`
+	Email  string `json:"email"`
+	Admin  bool   `json:"admin"`
+	//OnDemandTLS `json:",squash"`
+}
+
+// onConfigMapAdded is run when a config map is added to the cluster.
+// TODO It would be nice to filter configmaps by namespace directly
+//		on the informer
+func (c *CaddyController) onConfigMapAdded(obj interface{}) {
+	c.syncQueue.Add(ConfigMapAddedAction{
+		resource: obj,
+	})
+}
+
+// onConfigMapUpdated is run when an ingress resource is updated in the cluster.
+func (c *CaddyController) onConfigMapUpdated(old interface{}, new interface{}) {
+	c.syncQueue.Add(ConfigMapUpdatedAction{
+		resource:    new,
+		oldResource: old,
+	})
+}
+
+// onConfigMapDeleted is run when an ingress resource is deleted from the cluster.
+func (c *CaddyController) onConfigMapDeleted(obj interface{}) {
+	c.syncQueue.Add(ConfigMapDeletedAction{
+		resource: obj,
+	})
+}
+
+// ConfigMapAddedAction provides an implementation of the action interface.
+type ConfigMapAddedAction struct {
+	resource interface{}
+}
+
+// ConfigMapUpdatedAction provides an implementation of the action interface.
+type ConfigMapUpdatedAction struct {
+	resource    interface{}
+	oldResource interface{}
+}
+
+// ConfigMapDeletedAction provides an implementation of the action interface.
+type ConfigMapDeletedAction struct {
+	resource interface{}
+}
+
+func (r ConfigMapAddedAction) handle(c *CaddyController) error {
+	cfgMap, ok := r.resource.(*v1.ConfigMap)
+	if !ok {
+		return fmt.Errorf("ConfigMapAddedAction: incoming resource is not of type configmap")
+	}
+
+	// only care about the caddy config map
+	if !changeTriggerUpdate(c, cfgMap) {
+		return nil
+	}
+
+	logrus.Info("New configmap detected, updating Caddy config...")
+	// save to the store the current config map to use
+	c.resourceStore.ConfigMap = cfgMap
+
+	err := regenerateConfig(c)
+	if err != nil {
+		return err
+	}
+
+	logrus.Info("Caddy reloaded successfully.")
+	return nil
+}
+
+func (r ConfigMapUpdatedAction) handle(c *CaddyController) error {
+	cfgMap, ok := r.resource.(*v1.ConfigMap)
+	if !ok {
+		return fmt.Errorf("ConfigMapUpdatedAction: incoming resource is not of type configmap")
+	}
+
+	// only care about the caddy config map
+	if !changeTriggerUpdate(c, cfgMap) {
+		return nil
+	}
+
+	logrus.Info("ConfigMap resource updated, updating Caddy config...")
+
+	// save to the store the current config map to use
+	c.resourceStore.ConfigMap = cfgMap
+
+	err := regenerateConfig(c)
+	if err != nil {
+		return err
+	}
+
+	logrus.Info("Caddy reloaded successfully.")
+	return nil
+}
+
+func (r ConfigMapDeletedAction) handle(c *CaddyController) error {
```

Comment:
I?m not sure how to handle the configmap deletion when we define in the controller to use the configmap, than we receive an event that delete the configmap, I think we have two options, the first one, that's more safe approach, is to handle it by deleting the configuration, the second is to generate some more critical error and -maybe- kill the ingress controller.
The implemented solution seems "safe" but can be less clear to user that receive an error when we start the controller indicating that the configmap is needed but no configmap is available, but no error when we delete the configmap after the controller that declare to use the configmap is already started.

Any opinions?

Reply:
With the current implementation, if the config map does not exist at launch, it will log a warning and load without it. Maybe it's not the best idea and we should kill the process 🤔 

For now, we are "safe" and we do not require a ConfigMap to work. It may be interesting to have a way to make it required and when deleted, kill the process?

Reply:
Maybe we can start easy and -for example- panic if we need a CM and is not present (or someone delete it)